rfcs/012-add-session-ssh-support/012-add-session-ssh-support.md

```diff
@@ -0,0 +1,438 @@
+  development/support of new entities in the Renku platform
+
+A table showing how the different approaches compare is shown below:
+
+|                                                | Ssh port exposed by session directly | Proxy jumphost    | MITM Proxy         | OAuth2/OIDC       |
+|------------------------------------------------|--------------------------------------|-------------------|--------------------|-------------------|
+| VS Code support                                | :heavy_check_mark:                   | :heavy_check_mark: | :heavy_check_mark: | :x: |
+| Efficient IP address usage                     | :x:                                  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
+| Keys linked to user rather than project        | :x:                                  | :x:                | :heavy_check_mark:  | N/A |
+| Flexible monitoring support                    | :x:                                  | :x:                | :heavy_check_mark:  | :heavy_check_mark:                |
+| Key management responsibilities with users     | :heavy_check_mark:                   | :heavy_check_mark: | :heavy_check_mark:  | :x:                |
+| New Renku components to be developed/supported | :x:                                  | :x:                | :heavy_check_mark:  | :heavy_check_mark: |
+
+Based on the above comparison, the MITM proxy has the best fit with the requirements
+even though it requires the development of a dedicated, if modest component as
+part of Renku; it is also worth reiterating that this is the solution to which
+gitpod converged. 
+
+## Proposed Solution
+
+The proposed solution is based on the MITM Proxy described above. At an
+architectural level, this involves the addition of a component to the Renku
+platform which will terminate ssh sessions and route them to the appropriate
+user session. It will necessitate a solution for mapping between username, 
+keys used for login and the user session.
+
+### Mapping between ssh login and Renku user session
+
+Given that multiple Renku sessions could potentially use the same (user) ssh
+public key, it is not possible to use the ssh key itself as the mechanism
+which maps the ssh session to the Renku user session; this means it is not
+possible to have a solution in which the login is `ssh renku@<ssh-proxy>` and
+the session is uniquely identified by the ssh key. The most straightforward
+solution is one in which the username maps uniquely to a Renku user session.
+
+A mechanism is then required which maps ssh username to a Renku user session
+and the ssh challenge is successful using one of the user's registered ssh
+public keys.
+
+### Session creation
+
+The standard approach to session creation occurs in which the Renku notebooks
+service creates a `JupyterServer` CRD.
+
+In this case, however, a Kubernetes watcher listens for creation and deletion
+of these CRDs and updates a database accordingly. This database keeps track of
```

Comment:
I had to read this a few times to understand that this is actually part of the proposed change, i.e. that currently no such database exists. Maybe make that a bit more explicit.